[WIP] Autogenerate k8s manifests in skaffold init

nkubala opened the pull request:
*NOTE: this feature is hidden behind a feature flag, `--XXenableManifestGeneration`, to allow for further iteration and improvement before exposing publicly.*

this change adds support for autogenerating k8s manifests in `skaffold init` for projects that do not have any. skaffold will generate a k8s yaml for each found image that is not referenced by an existing k8s yaml in the project directory, and each generated yaml file will contain a deployment for the image, and a service exposing the deployment.

in the event that `skaffold init` finds a builder configuration in a project directory, but no k8s yaml referencing this image, we don't actually have any name for the image, so skaffold will generate a name. this name will either be:
1) the name of the encompassing directory, in the event that the builder is found in a nested directory; or,
2) the name of the builder file itself, in the event that only one builder configuration is found

for each image, a yaml file will be generated containing kubernetes resource definitions for a service and a deployment. this file will either be:
1) `<directory_name>/deployment.yaml`, for each builder that is found in a nested directory; or,
2) `<image_name>-deployment.yaml`, if we only have one builder configuration found

so for the `microservices` example, we would generate
```
.
├── leeroy-app
│   ├── app.go
│   ├── deployment.yaml <-- generated
│   └── Dockerfile
├── leeroy-web
│   ├── deployment.yaml <-- generated
│   ├── Dockerfile
│   └── web.go
├── README.md
└── skaffold.yaml <-- (also generated)
```

whereas for the `getting-started` example, we would generate
```
.
├── deployment.yaml <-- generated
├── Dockerfile
├── main.go
├── README.md
└── skaffold.yaml <-- generated
```

### Future Work
* figure out how to handle ports in the generated yaml - this is currently unsupported, so port forwarding will not work
* figure out the right behavior for resolved **and** unresolved builder configurations - if we have a preexisting `image1` being referenced in `k8s.yaml`, but also another builder configuration, should we always generate?
* remove feature flag

balopat commented on pkg/skaffold/deploy/deploy_mux_test.go:
```diff
@@ -57,9 +57,12 @@ func (m *MockDeployer) Render(_ context.Context, w io.Writer, _ []build.Artifact
-func (m *MockDeployer) WithDependenciesErr(err error) *MockDeployer      { m.dependenciesErr = err; return m }
-func (m *MockDeployer) WithCleanupErr(err error) *MockDeployer           { m.cleanupErr = err; return m }
-func (m *MockDeployer) WithRenderErr(err error) *MockDeployer            { m.renderErr = err; return m }
+func (m *MockDeployer) WithDependenciesErr(err error) *MockDeployer {
```
not sure why we have this in this PR.

nkubala replied:
the linter did this automatically for me. i'll take a look

nkubala commented:
@balopat whoops, in my haste to send this out for initial feedback I forgot to put a [WIP] in the title :laughing: it's in there now. this is definitely still WIP though it's taking the shape I want it to for an initial merge.

I also accidentally broke the microservices example when I was fixing an issue with the linter and pushed it right before opening this PR. it should be fixed now, can you give it a try?

> I don't think we should advertise a hidden flag in an error message ("if you'd like skaffold to generate a manifest for you, set the --XXenableManifestGeneration flag")

yeah I agree, I meant to remove this. in the future we'll advertise this here, but as long as the flag is hidden we shouldn't mention it. fixed

> if --skip-deploy is passed in or --kubernetes-manifests, we should not generate manifests - instead skaffold init --XXenableManifestGeneration --kubernetes-manifest hello.yaml results in a single prompt, then a single writing manifest - doesn't work

I *think* I agree with this, although I'm a little skeptical of the current behavior - it feels weird to me that passing `--kubernetes-manifest` to init doesn't process any artifacts into the generated skaffold.yaml by default. but if `--skip-deploy` is passed I agree we should not do anything here. I updated this to be in line with the current behavior, we can revisit this later if necessary.

> we need much more tests around this new functionality
:)

> we should break up this PR to separate out refactoring from functional changes
which parts are you talking about specifically? there isn't much refactoring in this PR...one thing got changed by the linter automatically which I will remove from this PR, but otherwise it's mostly new additions.

balopat commented on pkg/skaffold/initializer/builders.go:
```diff
@@ -206,28 +204,43 @@ func processCliArtifacts(artifacts []string) ([]builderImagePair, error) {
 	}
 
 	// if we only have 1 image and 1 build config, don't bother prompting
 	if len(images) == 1 && len(builderConfigs) == 1 {
```
you can remove this whole if - you basically copy pasted it to L228

balopat commented on pkg/skaffold/initializer/init.go:
```diff
@@ -71,6 +76,11 @@ type builderImagePair struct {
 	ImageName string
 }
 
+type unresolvedBuilderPair struct {
```
this name just doesn't add up to me, I don't understand it.

balopat commented on pkg/skaffold/initializer/init.go:
```diff
@@ -126,14 +134,43 @@ func DoInit(ctx context.Context, out io.Writer, c Config) error {
-			resolved, err := resolveBuilderImages(unresolvedBuilderConfigs, unresolvedImages, c.Force)
-			if err != nil {
+			resolved, unresolved, err := resolveBuilderImages(unresolvedBuilderConfigs, unresolvedImages, c.Force)
+			switch {
```
this whole logic here makes DoInit, the hardest to test part of `skaffold init` 30 lines longer. 
It is a nested switch inside an else ... with more ifs and for loops in there. 
This is not readable code.

nkubala replied:
as we discussed out of band, this PR was meant to open a discussion primarily on the UX/decision tree of the feature. I should have been more clear about that! I agree that this entire function is a mess and needs to be refactored/rewritten.

balopat commented:
pkg/skaffold/initializer/deploy_init.go | 57.14% <22.22%> (-26.2%) | ⬇️|
pkg/skaffold/initializer/init.go | 54.54% <40%> (-10.27%) | ⬇️|
pkg/skaffold/initializer/kubectl.go | 89.15% <62.5%> (-3.25%) | ⬇️|
pkg/skaffold/initializer/builders.go | 89.75% <72.22%> (-4.57%) | ⬇️|
